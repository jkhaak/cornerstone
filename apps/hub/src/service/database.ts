import pgLib from "pg-promise";

const pgp = pgLib({});

// eslint-disable-next-line dot-notation
const cn = process.env["DB_CN"];

if (!cn) {
  throw new Error('Environment variable "DB_CN" has not been set!');
}

const db = pgp(cn);

export { pgp, db };
