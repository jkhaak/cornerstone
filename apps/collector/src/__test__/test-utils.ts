import { db } from "../database";
import type { Response } from "supertest";
import { logger } from "@cornerstone/core";

/**
 * Teardown database connection after test run. Should be called in afterAll stage
 * outside of describe block.
 */
export function teardownTestConnection() {
  if (!db.$pool.ended) {
    db.$pool
      .end()
      .then(() => logger.info({ message: "DB connection closed succesfully" }))
      .catch((error: unknown) =>
        logger.error({ message: "Unexpected error happened while closing database connection.", error })
      );
  }
}

/**
 * Truncates given tables with cascade.
 */
export async function truncateTables(tables: string[]) {
  await db.tx((tx) => Promise.allSettled(tables.map((t) => tx.none("truncate table $1:name cascade", t))));
}

/**
 * Convert superttest headers from any to known record.
 */
export function getHeaders(response: Response): Record<Lowercase<string>, string> {
  const unknownHeaders: unknown = response.headers;

  if (Array.isArray(unknownHeaders)) {
    throw new Error("expected record not array");
  }

  if (
    Object.getOwnPropertySymbols(unknownHeaders).length > 0 &&
    !Object.getOwnPropertyNames(unknownHeaders).every((prop) => typeof prop === "string")
  ) {
    throw new Error("should have properties and property names should be string");
  }

  const headers = unknownHeaders as Record<string, unknown>;

  if (!Object.getOwnPropertyNames(headers).every((prop) => typeof headers[prop] === "string")) {
    throw new Error("properties should be string");
  }

  // return headers as Record<string, string>;
  return Object.fromEntries(
    Object.entries(headers as Record<string, string>).map(([prop, val]) => [prop.toLowerCase(), val])
  );
}
