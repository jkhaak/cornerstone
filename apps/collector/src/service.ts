import { db } from "./database";
import type { RuuviId } from "./model";

const SQL_INSERT_RUUVITAG = `
INSERT INTO ruuvitag
(id, mac)
VALUES($1, $2)
RETURNING id;
`;

const SQL_INSERT_RUUVIEVENT = `
INSERT INTO public.ruuvidata
(ruuvitag, "version", datetime, temperature, humidity, pressure, acceleration_x, acceleration_y, acceleration_z, power_voltage, power_tx, movement_counter, measurement_counter)
VALUES($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13);
`;

export async function saveEvent(event: unknown): Promise<RuuviId> {
  return `1234`;
}

export async function getEvent(id: RuuviId) {
  return {};
}
