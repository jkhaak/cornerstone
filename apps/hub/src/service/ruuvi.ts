import { RuuviData, ruuvi } from "@cornerstone/ruuvi-parser";
import { db } from "./database.js";
import { PreparedStatement as PS } from "pg-promise";

const SQL_INSERT_RUUVITAG = new PS({
  name: "insert-ruuvitag",
  text: `INSERT INTO ruuvitag (id, mac) VALUES ($1, $2) ON CONFLICT DO NOTHING;`,
});

const SQL_SELECT_RUUVITAGS = new PS({
  name: "select-ruuvitags",
  text: `SELECT id, mac FROM ruuvitag;`,
});

type RuuviEventDto = RuuviData & {
  tagId: string;
};

type RuuviTag = {
  tagId: string;
  mac: string;
};

export async function decodeEvent(eventBuffer: Buffer): Promise<RuuviEventDto> {
  const event = await ruuvi.decodeAsync(eventBuffer);

  return {
    ...event,
    tagId: event.mac.slice(-4),
  };
}

export async function fetchTags(): Promise<RuuviTag[]> {
  return db.manyOrNone(SQL_SELECT_RUUVITAGS);
}

export async function storeEvent(eventBuffer: Buffer) {
  const event = await decodeEvent(eventBuffer);

  await db.none(SQL_INSERT_RUUVITAG, [event.tagId, event.mac]);
}
