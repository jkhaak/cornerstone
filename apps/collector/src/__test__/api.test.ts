import request from "supertest";
import app from "../server";

import * as service from "../service";

import { rawData, rawEvent, ruuviTables } from "./ruuvi-service.test";
import { getHeaders, teardownTestConnection, truncateTables } from "./test-utils";
import _ from "lodash";
import type { RuuviData, RuuviId } from "../model";

function compareNumbers(a: number, b: number) {
  return a - b;
}

afterAll(() => {
  teardownTestConnection();
});

describe("rest api", () => {
  describe("get endpoints", () => {
    const ids = ["DEAD", "BEEF", "DADA"] as RuuviId[];
    const sequenceStart = _.random(20, 150);
    const measurementSequences = _.range(sequenceStart, sequenceStart + _.random(2, 7));

    beforeAll(async () => {
      await truncateTables(ruuviTables);

      await Promise.allSettled(
        ids.flatMap((id) =>
          measurementSequences.map((measurementSequence) =>
            service.saveEvent({
              ...rawEvent,
              ruuviId: id,
              data: { ...rawData, measurementSequence },
            })
          )
        )
      );
    });

    afterAll(async () => {
      await truncateTables(ruuviTables);
    });

    it("should return all known ruuvi tags", async () => {
      const response = await request(app).get("/ruuvi/tags");

      const headers = getHeaders(response);

      expect(headers["content-type"]).toMatch(/json/);
      expect(response.status).toBe(200);

      const idsSorted = [...ids].sort();
      const tagIdsSorted = [...(response.body as string[])].sort();

      expect(tagIdsSorted).toMatchObject(idsSorted);
    });

    it("should return all events by ruuvi tag", async () => {
      const id = ids[_.random(ids.length - 1)];
      if (id === undefined) {
        throw new Error("Couldn't pick any id from a test set.");
      }

      const response = await request(app).get(`/ruuvi/${id}/events`);
      expect(response.status).toBe(200);

      const headers = getHeaders(response);
      expect(headers["content-type"]).toMatch(/json/);

      const responseEvents = response.body as RuuviData[];
      const responseMeasurementSequences = responseEvents.map((event) => event.measurementSequence);
      responseMeasurementSequences.sort(compareNumbers);

      expect(responseMeasurementSequences).toMatchObject(measurementSequences);
    });

    it("should fail fetching events if requested with an unknown ruuvi id", async () => {
      const incorrectId = "TRSH";

      const response = await request(app).get(`/ruuvi/${incorrectId}/events`);
      expect(response.status).toBe(404);
      expect(response.body).toMatchObject({ error: `Couldn't find events with id ${incorrectId}` });
    });
  });

  describe("post endpoints", () => {
    afterEach(async () => {
      await truncateTables(ruuviTables);
    });

    it("should return resource id with created status", async () => {
      const id = "DEAF";
      const response = await request(app)
        .post("/ruuvi/event")
        .send({ ...rawEvent, data: { ...rawData, mac: `DEADBEEF${id}` } });
      expect(response.status).toBe(201);
      expect(response.body).toMatchObject({ id });
    });

    it("should consume ruuvi events", async () => {
      const ruuviId = "DEFA";
      const payload = { ...rawEvent, ruuviId, data: { ...rawData, mac: `DEADBEEF${ruuviId}` } };
      const response = await request(app).post("/ruuvi/event").send(payload);
      expect(response.status).toBe(201);

      const events = await service.getEvents(ruuviId);
      expect(events.length).toBe(1);
    });

    it("should prevent duplicate events", async () => {
      const firstPostResponse = await request(app)
        .post("/ruuvi/event")
        .send({ ...rawEvent, datetime: new Date() });
      expect(firstPostResponse.status).toBe(201);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const { id }: { id: RuuviId } = firstPostResponse.body;

      const secondPostResponse = await request(app)
        .post("/ruuvi/event")
        .send({ ...rawEvent, datetime: new Date() });
      expect(secondPostResponse.status).toBe(200);
      expect(secondPostResponse.body).toMatchObject({ id });

      const events = await service.getEvents(id);
      expect(events.length).toBe(1);
    });

    it("should consume multiple events (if they differ enough)", async () => {
      const firstPostResponse = await request(app)
        .post("/ruuvi/event")
        .send({ ...rawEvent, datetime: new Date(), data: { ...rawData, temperature: 24 } });
      expect(firstPostResponse.status).toBe(201);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const { id }: { id: RuuviId } = firstPostResponse.body;

      const secondPostResponse = await request(app)
        .post("/ruuvi/event")
        .send({ ...rawEvent, datetime: new Date(), data: { ...rawData, temperature: 25 } });
      expect(secondPostResponse.status).toBe(201);
      expect(secondPostResponse.body).toMatchObject({ id });

      const events = await service.getEvents(id);
      expect(events.length).toBe(2);
    });

    it("should fail with invalid event", () => {
      const invalidEvents = [
        {},
        { id: "DADA", datetime: new Date(), manufacturerDataHex: "123456789ABCDEF", data: {} },
        {
          id: "DADA",
          datetime: new Date(),
          manufacturerDataHex: "123456789ABCDEF",
          data: {
            manufacturerId: "449",
            version: 5,
            power: { tx: 5 },
          },
        },
      ];

      const expectedObject = {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        errorMessage: expect.stringMatching(/Validation error/),
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        issues: expect.anything(),
      };

      return Promise.all(
        invalidEvents.map(async (invalidEvent) => {
          const response = await request(app).post("/ruuvi/event").send(invalidEvent);
          expect(response.status).toBe(400);

          expect(response.body).toMatchObject(expectedObject);
          expect(response.body).toMatchSnapshot();
        })
      );
    });
  });
});
