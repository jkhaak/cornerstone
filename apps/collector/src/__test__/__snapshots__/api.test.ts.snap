// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rest api post endpoints should fail with invalid event 1`] = `
Object {
  "errorMessage": "Validation error: Required at \\"manufacturerDataBase64\\"",
  "issues": Array [
    Object {
      "code": "invalid_type",
      "expected": "string",
      "message": "Required",
      "path": Array [
        "manufacturerDataBase64",
      ],
      "received": "undefined",
    },
  ],
}
`;

exports[`rest api post endpoints should fail with invalid event 2`] = `
Object {
  "errorMessage": "Validation error",
}
`;

exports[`rest api post endpoints should fail with invalid event 3`] = `
Object {
  "errorMessage": "Validation error",
}
`;
